tokio/src/runtime/task/harness.rs

```diff
@@ -320,13 +340,33 @@ where
             if !snapshot.is_join_interested() {
                 // The `JoinHandle` is not interested in the output of
                 // this task. It is our responsibility to drop the
-                // output.
+                // output. The join waker was already dropped by the
+                // `JoinHandle` before.
                 self.core().drop_future_or_output();
             } else if snapshot.is_join_waker_set() {
                 // Notify the waker. Reading the waker field is safe per rule 4
                 // in task/mod.rs, since the JOIN_WAKER bit is set and the call
                 // to transition_to_complete() above set the COMPLETE bit.
                 self.trailer().wake_join();
+
+                // If JOIN_INTEREST is still set at this point the `JoinHandle`
+                // was not dropped since setting COMPLETE so we unset JOIN_WAKER
+                // to give the responsibility of dropping the join waker back to
+                // the `JoinHandle`. `JoinHandle` is able to drop the waker when
+                // itself gets dropped.
+                if self.state().unset_waker_if_join_interested().is_err() {
+                    // Unsetting JOIN_WAKER flag will fail if JOIN_INTERESTED is
+                    // not set to indicate that the runtime has the responsibility
+                    // to drop the join waker here as per rule 7 in task/mod.rs.
+                    // Safety:
+                    // If JOIN_INTEREST got unset since setting COMPLETE we are
+                    // the only ones to have access to the join waker and need
+                    // to drop it here because the `JoinHandle` of the task
+                    // already got dropped.
+                    unsafe {
+                        self.trailer().set_waker(None);
+                    }
+                }
```

Comment:
```suggestion
                // Inform the `JoinHandle` that we are done waking the waker by
		// unsetting the `JOIN_WAKER` bit. If the `JoinHandle` has
		// already been dropped, then this fails and we must drop the
		// waker ourselves.
                if self.state().unset_waker_if_join_interested().is_err() {
                    // SAFETY: We have COMPLETE=1 and JOIN_INTEREST=0, so
                    // we have exclusive access to the waker.
                    unsafe { self.trailer().set_waker(None) };
                }
```

Reply:
In principle this code could unset `JOIN_WAKER` unconditionally, which would allow us to use a `fetch_and` operation which is cheaper than a `compare_exchange` loop. After all, if the `JoinHandle` is already dropped, we don't care about the value of `JOIN_WAKER` after this call.

Reply:
You are right.
I made the change but I want to mention one thing: Now `State` basically has two functions to unset `JOIN_WAKER`: One that asserts the JOIN_INTEREST is set and the new one that asserts that the function is complete. I think it still makes sense to keep both of them but maybe the naming of the new one needs improvement? Right now I called it `unset_waker_after_complete`.